### PR TITLE
Adds `image_features` parameter to predict for pre-computed embeddings

### PR DESCRIPTION
**Pull request by NetZissou**
**Disclaimer:** This PR was developed with assistance from `Claude Opus 4.6 (1M context)`. The author has reviewed all code changes and test additions. CI has been executed successfully in the [forked repo](https://github.com/NetZissou/pybioclip/actions). Opening this PR to request review from the package maintainers for further feedback and iteration.

## Summary

This PR adds an optional `image_features` parameter to `predict()` on `TreeOfLifeClassifier` and `CustomLabelsClassifier` (`CustomLabelsBinningClassifier` inherits this through `CustomLabelsClassifier`). When provided, the method skips image encoding and computes classification directly from pre-computed embeddings.

## Embedding validation

The method validates input embeddings before classification:
- Verifies tensor is 2D `(N, embedding_dim)`
- Checks that `embedding_dim` matches the model's expected dimension (`model.visual.output_dim`)
- Normalizes the embedding vector via L2 norm only if not already normalized

## Test plan

New tests in `TestPredictFromEmbeddings`:
- Results from embeddings match results from images exactly (all fields except `file_name`)
- Species, family, and multi-image predictions
- Unnormalized features auto-normalized with correct classifications
- `CustomLabelsClassifier` and `CustomLabelsBinningClassifier`
- Error cases: no inputs, wrong tensor dim, wrong embedding dim, image-embedding length mismatch

Closes #167

### CODE REVIEW

**Review by hlapp — CHANGES_REQUESTED** [review]
Thanks @NetZissou. The part of the implementation approach that I don't like here is that now creating probabilities is taking place redundantly in two different functions. This also creates more code noise than I think should be needed in the `predict()` method. 

Instead, shouldn't the clean way to handle this in the `predict()` method be to see whether `image_features` are already provided. If they are, apply basic checks like correct dimensions etc. If they are not, create them (like the are being created now from images). Then proceed with creating probabilities from `image_features`.

**Comment by hlapp** [comment]
@NetZissou just FYI, it might be advisable to rebase on main to bring in the changes from #179. It's well possible your changes so far are not in conflict at all, but some stuff did get moved around.

File: src/bioclip/predict.py [hunk 1]
```diff
@@ -239,9 +239,99 @@ def create_image_features_for_image(self, image: str | PIL.Image.Image, normaliz
...
+        
+        # Image features provided
+        if image_features is not None:
+            # First validate dim of provided image_features
+            image_features = self._validate_image_features(image_features)
+            n = image_features.shape[0]
+            
+            # Check if images were provided
+            # If so, validate length and create keys; 
+            # if not, create dummy keys and empty images list for record_event
```
Comment (NetZissou) [inline src/bioclip/predict.py]: When `predict(image_features=feats, rank=...)` is called without images, `_resolve_image_features` returns `images=[]`. 

That empty list flows through `record_event` -> `PredictionRecorder.add_prediction`. 

The kwargs(`rank`, `min_prob`, `k`, `batch_size`) all record correctly, but `images=[]` carries no signal of how many predictions actually ran. 

Example output:
```
  Rank: species
  Min Prob: 1e-09
  K: 5
  Batch Size: 10
  Images:
```

Comment (NetZissou) [reply src/bioclip/predict.py]: Option: add `n_features` to `record_event` 

Pass `n_features=image_features.shape[0]` from `predict()` when the features path is used.

File: src/bioclip/predict.py [hunk 2]
```diff
@@ -239,9 +239,99 @@ def create_image_features_for_image(self, image: str | PIL.Image.Image, normaliz
...
+            if images is not None:
```
Comment (NetZissou) [inline src/bioclip/predict.py]: Currently `_resolve_image_features` accepts both `images` and `image_features` at the same time. The internal behavior in that case is:

- validate `image_features` and require it to match the length of `images`,
- use `image_features` to compute probabilities against text embeddings downstream,
- use `images` to construct cache keys.

After discussion with @egrace479 , this middle-ground case (both provided) seems to overcomplicate things. We think it would be cleaner to make `image_features` a drop-in replacement for `images` whenever it's provided, and never accept both.

One knock-on effect: `record_event` would need to change `images` from a required positional parameter to something optional, so we can still record an event when only `image_features` is provided.

@hlapp would love to hear your thoughts on the direction before we go further on the implementation.

Comment (hlapp) [reply src/bioclip/predict.py]: `images` can have metadata (such as name and/or path - watch out, for example, when the accepted type includes string) used for some things (constructing cache keys, for example), and presumably `image_features` would not.

Also, a vector representation of an image can be turned back into a displayable image, whereas a vector encoding (i.e., `image_features`) cannot. I don't think this difference matters anywhere right now (except possibly if one were debugging), so maybe that's more on the theoretical side.

Comment (NetZissou) [reply src/bioclip/predict.py]: Agreed. Current accept types for images are: string | a list of string | PIL.Image which can all be used to make keys. `image_features` is vector embeddings output from encoder, therefore does not contain any metadata information that could be used for key-gen.

Comment (egrace479) [reply src/bioclip/predict.py]: @NetZissou, @hlapp and I discussed the implementation this morning. It would make the most sense to expect (require) the list of image identifiers (e.g., list of filename strings) that correspond to the embeddings to be passed as well. These can then be used for the image keys. The embeddings would go through the size checks previously discussed and you have the sanity check that the list lengths match. We just want to make sure it's well-documented then.

**Comment by NetZissou** [comment]
@egrace479 @hlapp 

> @NetZissou, @hlapp and I discussed the implementation this morning. It would make the most sense to expect (require) the list of image identifiers (e.g., list of filename strings) that correspond to the embeddings to be passed as well. These can then be used for the image keys. The embeddings would go through the size checks previously discussed and you have the sanity check that the list lengths match. We just want to make sure it's well-documented then.

Re-implemented based on the comment above. Ready for re-review.

**Review by hlapp — CHANGES_REQUESTED** [review]
Thanks @NetZissou. Some thoughts and comments:

- Since this is trying to optimize performance for the case that someone has the image embeddings already, I think we can expect at least by default that they are already normalized, rather than re-normalizing them (for testing whether they are normalized) each time. So I think that should be conditional to a parameter that by default is off?
- The `_resolve_image_features()` function doesn't really resolve image features, given that it returns probabilities, not resolved image features. So instead it should probably better be named something like `_create_probabilities_for_imgs_or_img_feats()` (or maybe even just `_create_probabilities()`?). And given the number of parameters and `None` being allowed in several places, all invocations should arguably use named parameter syntax.
- In this implementation (as per `_resolve_image_features()`), there's actually really nothing directly in common between creating probabilities from image features and creating them from images: there is no code before or after the conditional that would run either way. I guess the shared part is indirect: creating probabilities from images invokes `create_batched_probabilities_for_images()`, whereas creating them from image features invokes `create_probabilities()`. Of course, indirectly the former invokes `create_probabilities_for_images()` which then invokes `create_probabilities()`, too. It seems what this really means is the following:
    * If I have images and no image features, encode the images to obtain image features. This should be done batched for performance optimization.
    * Once I have image features, create probabilities. This need not be batched.
  
  In your current implementation, creating probabilities from image features will be batched if starting from images, and the batching will be bypassed if image embeddings are already in hand. Is this reasonable?

**Comment by NetZissou** [comment]
> Since this is trying to optimize performance for the case that someone has the image embeddings already, I think we can expect at least by default that they are already normalized, rather than re-normalizing them (for testing whether they are normalized) each time. So I think that should be conditional to a parameter that by default is off?

I think it's worth adding an explicit param `normalize_features: bool`. By turning it off, user can assure that "I promise input vectors are already normalized". Regarding to the norm-check performance decrease, I think the main goal of this PR is to reduce the unnecessary image encoding (ViT forward pass).
```
N: number of entries
D: vector dimension
C: number of classes in the text embeddings

Dimensions: 
image_features: (N, D)
txt_embeddings: (D, C)
logits/probs: (N, C)

norm check: N x D
logits matmul: N x C x D

(norm check) / matmul = 1/C
```
The norm check costs about `1/C` of the logits matmul. For BioCLIP 2, norm check is about 10^5 cheaper than matmul. **And they both are dwarfed by image encoding**. I think we can afford to keep this cheap norm sanity check that warns or raises instead of silently producing wrong probs. Setting `normalize_features = False` bypass the norm check. 
 
---

> The `_resolve_image_features()` function doesn't really resolve image features, given that it returns probabilities, not resolved image features. So instead it should probably better be named something like `_create_probabilities_for_imgs_or_img_feats()` (or maybe even just `_create_probabilities()`?). And given the number of parameters and None being allowed in several places, all invocations should arguably use named parameter syntax.

Agreed. I prefer `_create_probabilities_for_imgs_or_img_feats()` because `_create_probabilities()` name would confuse with the existing [`create_probabilities()`](https://github.com/NetZissou/pybioclip/blob/1ae72a4f4b3d240dbcaf2a46c921f0024bee23f4/src/bioclip/predict.py#L240). 

---
> In your current implementation, creating probabilities from image features will be batched if starting from images, and the batching will be bypassed if image embeddings are already in hand. Is this reasonable?

You're right, the image features path should be batched! 

Current implementation:
- if only images, no image features provided, [the image encoding is batched `create_batched_probabilities_for_images`](https://github.com/NetZissou/pybioclip/blob/1ae72a4f4b3d240dbcaf2a46c921f0024bee23f4/src/bioclip/predict.py#L328)
- if images, and image features are both provided, [the logits/probs calculation `image_features @ txt_features` is not batched](https://github.com/NetZissou/pybioclip/blob/1ae72a4f4b3d240dbcaf2a46c921f0024bee23f4/src/bioclip/predict.py#L322), which can cause OOM errors for large `N`.

**Comment by hlapp** [comment]
> The norm check costs about `1/C` of the logits matmul. For BioCLIP 2, norm check is about 10^5 cheaper than matmul. **And they both are dwarfed by image encoding**. I think we can afford to keep this cheap norm sanity check that warns or raises instead of silently producing wrong probs.

Probably. But your argument assumes that norm check and matmul use the same hardware acceleration, which may or may not be true, and it's also just a matter of future-proofing principle.

**Comment by NetZissou** [comment]
@hlapp thanks, this makes sense. In [`_validate_image_features`](_validate_image_features), I'll drop unit norm check [`torch.allclose`](https://github.com/NetZissou/pybioclip/blob/1ae72a4f4b3d240dbcaf2a46c921f0024bee23f4/src/bioclip/predict.py#L281) entirely, still keeping the tensor dimension checks to be defensive, good for error catching. 

Introduce new parameter `normalize_features`:
- `True`: unconditionally `F.normalizes` the input
- `False (default)`: skip normalization entirely

**Comment by hlapp** [comment]
@NetZissou were you still going to implement the change as per above? Just wanted to make I didn't misunderstand your response/comment.

**Comment by NetZissou** [comment]
@hlapp yes that's still in the work
